pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -275,11 +327,40 @@ func (ctrl *PersistentVolumeController) shouldDelayBinding(claim *v1.PersistentV
 	if ctrl.isDelayBindingProvisioning(claim) {
 		return false, nil
 	}
-
 	// If claim is in delay binding mode.
 	return pvutil.IsDelayBindingMode(claim, ctrl.classLister)
 }
 
+// Unit test [12-2] [12-4]
+func createOperationTimestampCacheKey(opName, key string) string {
+	return key + "-" + opName
+}
+
+func (ctrl *PersistentVolumeController) recordMetric(opName, key string, err error) {
```

Comment:
Can we separate the entire metric collection concern inside a new type? I am thinking something like:

```golang
type ProvisionMetric struct {
  cache cache.ThreadSafeStore
  ....
}
```

and define this type in a new file. All the functions(as well as we can) that deal with metrics code in pv_controller could be moved to this new file. We should try and reduce coupling between this file and pv_controller file.

Reply:
Potentially in the future we can refactor the file to work for the other volume controllers too, similar to the one in operation generator

Reply:
Where I agree separating the logics into another file makes it cleaner and its really not a big effort for me to do it, my major concern is there are already two places doing it. One is persistentvolume/metrics/metrics.go, the other is util/metrics.go (both in this PR). Creating a new file in the current package will make it even more confusing. I thought about putting logics into persistentvolume/metrics/metrics.go, however it will end up opening up whole bunch of public apis from that file.

Reply:
I think those metrics logic need another effort to merge/consolidate.

Reply:
moved to persistentvolume/metrics/metrics.go per discussion